File: engine/source/tests/tainting/paintedwolf/perl-short-circuit/flow.pl
my $input = source();
0 && sink($input);
'0' && sink($input);
'' && sink($input);
undef && sink($input);
'yes' || sink($input);
1 || sink($input);
'0' // sink($input);
'' // sink($input);
# ruleid: flow
1 && sink($input);
# ruleid: flow
'yes' && sink($input);
# ruleid: flow
0 || sink($input);
# ruleid: flow
'' || sink($input);
# ruleid: flow
undef // sink($input);
0 and sink($input);
'yes' or sink($input);
# ruleid: flow
1 and sink($input);
# ruleid: flow
0 or sink($input);
my $effect = 'fixed';
0 && ($effect = source());
sink($effect);
my $alternate = 'fixed';
'yes' || ($alternate = source());
sink($alternate);
my $default = 'fixed';
'' // ($default = source());
sink($default);
my $chosen = undef // source();
# ruleid: flow
sink($chosen);
my $retained = source() // 'fixed';
# ruleid: flow
sink($retained);
my $safe = 'fixed' || source();
sink($safe);
my $empty = '' && source();
sink($empty);

File: engine/source/tests/tainting/paintedwolf/clojure-collections/flow.clj
(ns app (:require [clojure.core :as core]))
(defn collections [key]
  (let [original {:query (source) :safe "fixed"}
        revised (assoc original :query "fixed")]
    (sink (:query revised))
    ;; ruleid: flow
    (sink (:query original))
    (sink (get revised :safe))
    (sink (get revised :query)))
  (let [original {:query "fixed" :safe "fixed"}
        revised (assoc original :query (source))]
    ;; ruleid: flow
    (sink (get revised :query))
    (sink (get original :query))
    (sink (:safe revised)))
  (let [original [(source) "fixed"] revised (core/assoc original 0 "fixed")]
    (sink (core/get revised 0))
    ;; ruleid: flow
    (sink (core/get original 0)))
  (let [original {:query (source)} revised (clojure.core/assoc original :query "fixed")]
    (sink (clojure.core/get revised :query)))
  (let [original {:query "fixed"} revised (assoc original key (source))]
    ;; ruleid: flow
    (sink (:query revised)))
  (let [original {:query "fixed"} revised (assoc original :query (source) :query "fixed")]
    (sink (:query revised)))
  (let [assoc (fn [value key replacement] value)
        original {:query (source)} revised (assoc original :query "fixed")]
    ;; ruleid: flow
    (sink (:query revised)))
  (let [get (fn [value key] (:unsafe value))]
    ;; ruleid: flow
    (sink (get {:query "fixed" :unsafe (source)} :query))))

(defn unknown-collection []
  (let [original (source) revised (assoc original :query "fixed")]
    (sink (:query revised))
    ;; ruleid: flow
    (sink (:other revised))
    ;; ruleid: flow
    (sink (:query original))))

(defn nested-overwrite []
  (let [original (source) revised (assoc original :query {"input" "fixed"})]
    (sink (get (:query revised) "input"))))
